### PR TITLE
[Hack week] Notify about card reader optional update after every payment collection

### PR DESCRIPTION
**Pull request by AnirudhBhat**
### Description

This PR notifies (snackbar) users about optional card reader updates, if available, after every payment.

### Why?
Every card reader update starts out as an optional update. When the optional update is ignored for a certain time, it becomes a mandatory update and needs to be updated before collecting any payment. 

Mandatory updates will block users from taking payment. This will be a frustrating experience for users especially when there are customers waiting in front of them.

From our [track](https://mc.a8c.com/tracks/trends/?compareby=prop&eventname=woocommerceandroid_card_reader_software_update_started&interval=day&prop=software_update_type&startdate=20230829&stat=uniques) [data](https://mc.a8c.com/tracks/trends/?compareby=prop&eventname=woocommerceios_card_reader_software_update_started&interval=day&prop=software_update_type&startdate=20230829&stat=uniques), it looks like most of the update that happens are mandatory. This clearly indicates that the users aren't quite aware of the optional card reader update and need more nudges to install it while it is still an optional update.

In order to create awareness, we try to display an unintrusive message that informs users and allows them to install an optional card reader update if it is available on the payments hub screen.

### Impact
Merchants would install card reader updates when it is optional (ideally, when there are no customers standing in front of them) and this would avoid them having to install mandatory updates while collecting the payment.

#### If there is an optional card reader update available, make sure after every payment, you see a notification about the update.

1. In `TerminalWrapper.kt`, line 101 - Change `CardReaderManager.SimulatorUpdateFrequency.ALWAYS -> SimulateReaderUpdate.REQUIRED` to `CardReaderManager.SimulatorUpdateFrequency.ALWAYS -> SimulateReaderUpdate.UPDATE_AVAILABLE`
2. Navigate to the settings tab
3. Click on the Developer Options row
4. Enable simulated card reader
5. Select "Update simulated card reader" to `Always`
6. Navigate to the payments menu
7. Collect IPP
8. Make sure you see the snack bar about the optional update if there is one
9. Click on the `install` action button and ensure the update happens

#### If there is an optional card reader update available and the card reader is connected, you see a notification about the update every time on the payments hub.
1. In `TerminalWrapper.kt`, line 101 - Change `CardReaderManager.SimulatorUpdateFrequency.ALWAYS -> SimulateReaderUpdate.REQUIRED` to `CardReaderManager.SimulatorUpdateFrequency.ALWAYS -> SimulateReaderUpdate.UPDATE_AVAILABLE`
2. Navigate to the settings tab
3. Click on the Developer Options row
4. Enable simulated card reader
5. Select "Update simulated card reader" to `Always`
6. Navigate to the payments menu
7. Ensure you see the snack bar about the update

### Images/gif

https://github.com/woocommerce/woocommerce-android/assets/1331230/d5607a9c-1351-4492-bf6d-1063300f43be

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -1408,6 +1408,7 @@
     <string name="card_reader_payment_print_receipt">Print receipt</string>
     <string name="card_reader_payment_send_receipt">Send receipt</string>
     <string name="card_reader_payment_save_for_later">Save receipt and continue</string>
+    <string name="card_reader_payment_update_available">There\'s an update available for the card reader</string>
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/res/values/strings.xml]: I'm open to any suggestions here on better wording.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/UIMessageResolver.kt [hunk 2]
```diff
@@ -47,6 +47,27 @@ interface UIMessageResolver {
...
     }
 
+    /**
+     * Create and return a snackbar displaying the provided message and a Install button.
+     *
+     * @param [stringResId] The string resource id of the base message
+     * @param [stringArgs] Optional. One or more format argument stringArgs
+     * @param [actionListener] Listener to handle the undo button click event
+     */
+    fun getInstallSnack(
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/base/UIMessageResolver.kt:57]: If you'll go with `Indefinite` which I think we should, then you won't need this method, but you could use the one above `getRestartSnack` - just change the comment and maybe the name of the method

File: WooCommerce/src/main/res/values/strings.xml (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/res/values/strings.xml]: What about `An update for the card reader is available.` for the sake of shortness?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/hub/PaymentsHubViewModel.kt [hunk 3]
```diff
@@ -96,6 +101,36 @@ class PaymentsHubViewModel @Inject constructor(
...
+                            cardReaderTracker.trackSoftwareUpdateAlertInstallClicked()
+                            triggerEvent(PaymentsHubEvents.CardReaderUpdateScreen)
+                        }
+                    )
+                )
+            }
+            SoftwareUpdateAvailability.NotAvailable -> {
+                // no op
+            }
+        }.exhaustive
```
Comment (kidinov) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/hub/PaymentsHubViewModel.kt:131]: np: In the latest kotlin `exhaustive` is not needed as all "when" are expressions. We'll need to remove all `exhaustive`

**Review by kidinov — COMMENTED** [review]
LGTM!

A few thoughts:
* As I mentioned already I think having snack bar that needs action to be removed may be the right move here. Wdyt?
* Simple payments may be removed soon and the main place where all the payments will be collected will be order details. Even now most of the payments are done via order details. I am thinking maybe to show the snackbar right after the payment using the dialog context. This way that will be working everywhere I guess?
* And the blocking one - we'll need to add that to the release notes

**Comment by AnirudhBhat** [comment]
@kidinov 
> As I mentioned already I think having a snack bar that needs action to be removed may be the right move here. Wdyt?

The sticky snack bar will persist throughout the app until the merchants click the "Install" button or swipe to dismiss. While the idea of having a persistent snack bar that remains until the merchant takes action (either clicking "Install" or swiping to dismiss) might increase awareness of optional updates, there's a potential concern about user frustration. The swipe action to dismiss the snack bar may not be intuitive for all users, and there's a risk that many won't even realize the snack bar is dismissible.

Even though this might increase the number of optional updates, I'm afraid that this might cause frustration for some users as the notification might appear as a blocking one or as a distraction for them.

WDYT?

**Review by kidinov — APPROVED** [review]
@AnirudhBhat 

I see your concern, but [INDEFINITE](https://developer.android.com/reference/com/google/android/material/snackbar/BaseTransientBottomBar#LENGTH_INDEFINITE) is the official Google approach to snack bars, so it's not like we are hacking something here. That's very much the same as having dialog with one action button, and we are not afraid that people won't be able to figure out how to hide it.

At the end of the day, it's your call, and if you think that, it's better to have that finite, I am totally fine with that

**Comment by AnirudhBhat** [comment]
@kidinov 
> Simple payments may be removed soon and the main place where all the payments will be collected will be order details. Even now most of the payments are done via order details. I am thinking maybe to show the snackbar right after the payment using the dialog context. This way that will be working everywhere I guess?

This was my original thought as well. To implement another dialog that informs about the card reader update availability after the payment transaction finishes. For the HACK week, to keep it simple, I decided on the snack bar.

I'll think more about this and raise a separate PR implementing the said approach. If we go with this approach, we don't have to worry about the length of the snack bar notification as well.

Thank you!

**Comment by kidinov** [comment]
@AnirudhBhat 

> This was my original thought as well. To implement another dialog that informs about the card reader update availability after the payment transaction finishes. For the HACK week, to keep it simple, I decided on the snack bar.

What I meant is that you can do the same what you have done, but using "successful" state of the CardReaderPaymentsDialog and then the snack bar will be shown on both flows - simple payments and order

**Pull request merged**